pkg/controllers/management/rkeworkerupgrader/upgrade.go

```diff
@@ -0,0 +1,395 @@
+			return nil, err
+		}
+		return node, nil
+	}
+
+	// proceed only if node and cluster's versions mismatch
+	if cluster.Status.NodeVersion == node.Status.AppliedNodeVersion {
+		return node, nil
+	}
+
+	nodePlan, err := uh.getNodePlan(node, cluster)
+	if err != nil {
+		return nil, err
+	}
+
+	if node.Status.AppliedNodeVersion == 0 {
+		// node never received appliedNodeVersion
+		if planChangedForUpgrade(nodePlan, node.Status.NodePlan.Plan) || planChangedForUpdate(nodePlan, node.Status.NodePlan.Plan) {
+			return uh.updateNodePlan(node, cluster)
+		}
+	} else {
+		if planChangedForUpgrade(nodePlan, node.Status.NodePlan.Plan) {
+			logrus.Infof("cluster [%s] worker-upgrade: plan changed for node [%s], call upgrade to reconcile cluster", cluster.Name, node.Name)
+			if err := uh.upgradeCluster(cluster, node.Name); err != nil {
+				return nil, err
+			}
+			return node, nil
+		}
+		if planChangedForUpdate(nodePlan, node.Status.NodePlan.Plan) {
+			logrus.Infof("cluster [%s] worker-upgrade: plan changed for update [%s]", cluster.Name, node.Name)
+			return uh.updateNodePlan(node, cluster)
+		}
+	}
+
+	return node, nil
+}
+
+func (uh *upgradeHandler) updateNodePlan(node *v3.Node, cluster *v3.Cluster) (*v3.Node, error) {
+	nodePlan, err := uh.getNodePlan(node, cluster)
+	if err != nil {
+		return nil, fmt.Errorf("getNodePlan error for node [%s]: %v", node.Name, err)
+	}
+
+	nodeCopy := node.DeepCopy()
+	np := &v3.NodePlan{
+		Plan:    nodePlan,
+		Version: cluster.Status.NodeVersion,
+	}
+	if node.Status.NodePlan == nil || node.Status.NodePlan.AgentCheckInterval == 0 {
+		// default
+		np.AgentCheckInterval = nodeserver.DefaultAgentCheckInterval
+	} else {
+		np.AgentCheckInterval = node.Status.NodePlan.AgentCheckInterval
+	}
+	nodeCopy.Status.NodePlan = np
+
+	logrus.Infof("cluster [%s] worker-upgrade: updating node [%s] with plan [%v]", cluster.Name, node.Name, np.Version)
+
+	updated, err := uh.nodes.Update(nodeCopy)
+	if err != nil {
+		return nil, fmt.Errorf("error updating node [%s] with plan %v", node.Name, err)
+	}
+
+	return updated, err
+}
+
+func (uh *upgradeHandler) updateNodePlanVersion(node *v3.Node, cluster *v3.Cluster) (*v3.Node, error) {
+	nodeCopy := node.DeepCopy()
+	nodeCopy.Status.NodePlan.Version = cluster.Status.NodeVersion
+	logrus.Infof("cluster [%s] worker-upgrade: updating node [%s] with plan version [%v]", cluster.Name,
+		node.Name, nodeCopy.Status.NodePlan.Version)
+
+	updated, err := uh.nodes.Update(nodeCopy)
+	if err != nil {
+		return nil, err
+	}
+	return updated, err
+
+}
+
+func (uh *upgradeHandler) getNodePlan(node *v3.Node, cluster *v3.Cluster) (*v3.RKEConfigNodePlan, error) {
+	var (
+		nodePlan *v3.RKEConfigNodePlan
+		err      error
+	)
+	if isNonWorkerOnly(node.Status.NodeConfig.Role) {
+		nodePlan, err = uh.nonWorkerPlan(node, cluster)
+	} else {
+		nodePlan, err = uh.workerPlan(node, cluster)
+	}
+	return nodePlan, err
+}
+
+func lock(name string) {
```

Comment:
Looks like you just recreated https://github.com/moby/moby/tree/master/pkg/locker